crates/ingress-client/src/session.rs

```diff
@@ -0,0 +1,501 @@
+        let _ = self.tx.send(Err(CommitError::Cancelled));
+    }
+}
+
+struct IngressBatch {
+    records: Arc<[IngestRecord]>,
+    trackers: Vec<RecordTracker>,
+    reply_rx: Option<ReplyRx<IngestResponse>>,
+}
+
+impl IngressBatch {
+    /// Marks every tracked record in the batch as committed.
+    pub fn committed(self) {
+        for tracker in self.trackers {
+            tracker.resolver.committed();
+        }
+    }
+}
+
+/// Tunable parameters for batching and networking behaviour of partition sessions.
+#[derive(Debug, Clone)]
+pub struct SessionOptions {
+    /// Maximum batch size
+    pub batch_size: usize,
+    /// Maximum batch timeout before it's sent
+    pub batch_timeout: Duration,
+    /// connection retry policy
+    pub connect_retry_policy: RetryPolicy,
+}
+
+impl Default for SessionOptions {
+    fn default() -> Self {
+        Self {
+            batch_size: 250,
+            batch_timeout: Duration::from_millis(100),
+            connect_retry_policy: RetryPolicy::exponential(
+                Duration::from_millis(10),
+                2.0,
+                None,
+                Some(Duration::from_secs(1)),
+            ),
+        }
+    }
+}
+
+struct RecordTracker {
+    _permit: OwnedSemaphorePermit,
+    resolver: RecordCommitResolver,
+}
+
+/// Cloneable sender that enqueues records for a specific partition session.
+#[derive(Clone)]
+pub struct SessionHandle {
+    tx: mpsc::UnboundedSender<(RecordTracker, IngestRecord)>,
+}
+
+impl SessionHandle {
+    /// Enqueues an ingest request along with the owned permit and returns a future tracking commit outcome.
+    pub fn ingest(
+        &self,
+        permit: OwnedSemaphorePermit,
+        record: IngestRecord,
+    ) -> Result<RecordCommit, SessionClosed> {
+        let (commit, resolver) = RecordCommit::new();
+        self.tx
+            .send((
+                RecordTracker {
+                    _permit: permit,
+                    resolver,
+                },
+                record,
+            ))
+            .map_err(|_| SessionClosed)?;
+
+        Ok(commit)
+    }
+}
+
+enum SessionState {
+    Connecting { retry: RetryIter<'static> },
+    Connected { connection: Connection },
+    Disconnected,
+    Shutdown,
+}
+
+/// Background task that drives the lifecycle of a single partition connection.
+pub struct PartitionSession<T> {
+    partition: PartitionId,
+    partition_routing: PartitionRouting,
+    networking: Networking<T>,
+    opts: SessionOptions,
+    rx: UnboundedReceiverStream<(RecordTracker, IngestRecord)>,
+    tx: mpsc::UnboundedSender<(RecordTracker, IngestRecord)>,
+    inflight: VecDeque<IngressBatch>,
+}
+
+impl<T> PartitionSession<T> {
+    fn new(
+        networking: Networking<T>,
+        partition_routing: PartitionRouting,
+        partition: PartitionId,
+        opts: SessionOptions,
+    ) -> Self {
+        let (tx, rx) = mpsc::unbounded_channel();
+        let rx = UnboundedReceiverStream::new(rx);
```

Comment:
It's not yet fully clear to me how we will ensure that this stream does not grow unboundedly. I guess this will become clearer from the following PRs.

Reply:
Ideally, users could specify the window size in bytes that a session can use to pipeline sending events.

Reply:
There is now 2 knobs that can be used to control the client.
- A budget on the client, control how much data can be inflight without confirming (in bytes)
- A batch size (in bytes) how big a single request can be.